caching-service/src/main/java/org/zowe/apiml/caching/service/infinispan/storage/InfinispanStorage.java

```diff
@@ -153,4 +151,75 @@ public void deleteForService(String serviceId) {
             }
         });
     }
+
+    @Override
+    public void deleteItemFromMap(String serviceId, String mapKey) {
+        if (mapKey.equals("invalidUsers") || mapKey.equals("invalidScopes")) {
+            removeNonRelevantRules(serviceId, mapKey);
+        } else {
+            removeNonRelevantTokens(serviceId, mapKey);
+        }
+    }
+
+    private void removeNonRelevantTokens(String serviceId, String mapKey) {
+        Map<String, String> map = tokenCache.get(serviceId + mapKey);
```

Comment:
you need to obtain the lock(lock.tryLock) before calling the cache as it can be modified in the meantime and you would store the old version of the map in the cache

Reply:
Nice point, I will change it